### PR TITLE
Setup header header struct.

### PR DESCRIPTION
**Pull request by james-bird**
Very slow to construct/read into. I think it's to do with the use of static arrays, but haven't found a fix yet. Not yet tested the validity of data.

### CODE REVIEW

**Comment by stillyslalom** [comment]
Current header compilation time:
```
julia> t = time(); CineFiles.CineHeader("test/data/8bpp.cine"); time() - t
0.07999992370605469
```
After this PR:
```
julia> t = time(); CineFiles.CineHeader("test/data/8bpp.cine"); time() - t
46.08899998664856
```
Yeah, that's pretty painful. I think the main culprit is `Description::SVector{4096,Char}`, since compilation time drops to ~7.5 seconds if I comment that line out of the struct definition. Julia's compiler has issues with large tuples (https://github.com/JuliaLang/julia/issues/35619), and the solution probably involves restructuring the setup header into a non-`isbits` type with `String`s replacing each `NTuple{N,Char}`. I'll take a shot at it.

**Comment by james-bird** [comment]
The issue could be related to the issue discussed here:
https://discourse.julialang.org/t/long-parameter-for-type-seems-very-slow/63290

**Comment by james-bird** [comment]
I have made some changes which make the reading much more tolerable. The data looks sensible too, based on my limited testing.

File: src/CineFiles.jl [hunk 1]
```diff
@@ -2,19 +2,219 @@ module CineFiles
...
+    Res2::Int32
+    Res3::Int32
+    ImWidth::UInt16
+    ImHeight::UInt16
+    EDRShutter16::UInt16
+    Serial::UInt32
+    Saturation::Int32
+    Res5::UInt8
+    AutoExposure::UInt32
+    bFlipH::SizedVector{4,Bool}
```
Comment (stillyslalom) [inline src/CineFiles.jl:111]: Bools are aliased to Int32s in the header:
```
typedef int bool32_t;
```
...but Julia `Bool`s are one byte apiece. Can you swap these out for Int32s?

File: src/CineFiles.jl [hunk 2]
```diff
@@ -59,9 +258,13 @@ function CineHeader(fname)
     open(fname) do f
         read(f, UInt16) == UInt(18755) || error(basename(fname), " is not a .cine file")
         seek(f, 0)
-        cine = readstruct(f, CineFileHeader)
+        cine = read(f, CineFileHeader)
         seek(f, cine.OffImageHeader)
-        bitmap = readstruct(f, BitmapInfoHeader)
+        bitmap = read(f, BitmapInfoHeader)
+
+        seek(f, 0)
```
Comment (stillyslalom) [inline src/CineFiles.jl:264]: `seek` specifies a position, not an offset, so only one `seek` is needed

File: src/CineFiles.jl (hunk 1 above)
Comment (stillyslalom) [inline src/CineFiles.jl]: Bools are aliased to Int32s in the header:
```
typedef int bool32_t;
```
...but Julia `Bool`s are one byte apiece. Can you swap these out for Int32s?

File: src/CineFiles.jl [hunk 3]
```diff
@@ -2,19 +2,219 @@ module CineFiles
...
 export CineFile
 
 # using FileIO
 # add_format(format"CINE", [0x43, 0x49], ".cine")
 
-struct Time64
+abstract type BinaryData end
+
+read(f::IOStream, S::Type{T}) where {T<:BinaryData} = S(read.(Ref(f), S.types)...)
+read(f::AbstractString, S::Type{T}) where {T<:BinaryData} = S(read.(f, S.types)...)
```
Comment (stillyslalom) [inline src/CineFiles.jl]: What's the purpose of the `AbstractString` method? Why broadcast over the string?

**Review by stillyslalom — COMMENTED** [review]
Looks mostly good - I've also considered iteratively reading the setup struct as a `OrderedDict{Symbol,Any}`, which might fractionally increase runtime, but would sharply decrease compilation time. That would also make it easier to handle the branches necessary to read older `Setup` header formats.

File: src/CineFiles.jl (hunk 3 above)
Comment (james-bird) [reply src/CineFiles.jl]: That can go. Its just in there from when I was debugging something and forgot to remove it.

File: src/CineFiles.jl (hunk 1 above)
Comment (james-bird) [reply src/CineFiles.jl]: Yep, sounds good. Ideally it would be nice to have a 32 bit bool.

File: src/CineFiles.jl (hunk 2 above)
Comment (james-bird) [reply src/CineFiles.jl:264]: 👍

**Pull request closed without merging**